src/main/java/uk/ac/ebi/ega/egafuse/service/EgaFileService.java

```diff
@@ -86,19 +103,64 @@ public EgaFileService(OkHttpClient okHttpClient, String apiURL, Token token) {
                     filename = filename.substring(filename.lastIndexOf("/") + 1);
                 }
 
-                String type = "SOURCE";
                 if (filename.toLowerCase().endsWith(".gpg")) {
-                    type = "GPG";
                 } else if (filename.toLowerCase().endsWith(".cip")) {
-                    type = "CIP";
+                    file.setFileSize(file.getFileSize() - 16);
                 }
 
-                egaFiles.add(new EgaFile(filename.substring(0, filename.length() - 4), type, file));
+                egaFiles.add(new EgaFile(filename.substring(0, filename.length() - 4), file, this));
             }
             return egaFiles;
         default:
             LOGGER.error("status: {}", status);
             throw new ClientProtocolException(response.body().string());
         }
     }
+
+    public int fillBufferCurrentPage(Pointer buffer, String fileId, long totalFileSize, long currentPageSize,
+            long offset) {
+        int bytesToRead = (int) Math.min(totalFileSize - offset, currentPageSize);
+        int currentPage = (int) (offset / PAGE_SIZE);
+
+        if (offset >= totalFileSize || bytesToRead <= 0)
+            return -1;
+
+        loadPage(totalFileSize, fileId, currentPage, totalFileSize);
+
+        byte[] page = null;
+        final String key = fileId + "_" + currentPage;
+        try {
+            if (keys.get(key).get() != null) {
+                page = (byte[]) cachemanager.getCache("archive").get(key).get();
+            }
+        } catch (InterruptedException | ExecutionException e) {
+            LOGGER.error("Error in reading from cachemanager - {}", e.getMessage());
+        } 
+
+        if (page == null) {
+            LOGGER.error("Service seems to be down");
+            return -1;
+        } else {
+            int pageOffset = (int) (offset - currentPage * PAGE_SIZE);
+            buffer.put(0L, page, pageOffset, bytesToRead);
+            return bytesToRead;
+        }
+    }
+
+    private void loadPage(long fsize, String fileId, int currentPage, long fileSize) {
+        int maxPage = (int) (fsize / PAGE_SIZE);
+        int nextEndPage = currentPage + NUM_PAGES;
+        nextEndPage = nextEndPage > maxPage ? maxPage : nextEndPage;
+
+        for (; currentPage <= nextEndPage; currentPage++) {
+            final int nextPageNumber = currentPage;
+            final String key = fileId + "_" + nextPageNumber;
+            if (!keys.containsKey(key)) {
```

Comment:
You can use keys.computeIfAbsent(key, k -> {})

Reply:
Not achieved the same goal using computeIfAbsent() so didn't change it
(The value executor job is always run even if the key is present in the keys map)

Reply:
This should not happen. computeIfAbsent() expected to execute if key doesn't exists. We can check what's going wrong.

Reply:
Done